tokio/src/util/wake_list.rs

```diff
@@ -0,0 +1,47 @@
+use core::mem::MaybeUninit;
+use std::task::Waker;
+
+const NUM_WAKERS: usize = 32;
+
+pub(crate) struct WakeList {
+    inner: [MaybeUninit<Waker>; NUM_WAKERS],
+    curr: usize,
+}
+
+impl WakeList {
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: unsafe { MaybeUninit::uninit().assume_init() },
+            curr: 0,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn can_push(&self) -> bool {
+        self.curr < NUM_WAKERS
+    }
+
+    pub(crate) fn push(&mut self, val: Waker) {
+        debug_assert!(self.can_push());
+
+        self.inner[self.curr] = MaybeUninit::new(val);
+        self.curr += 1;
+    }
+
+    pub(crate) fn wake_all(&mut self) {
+        assert!(self.curr <= NUM_WAKERS);
+        while self.curr > 0 {
+            self.curr -= 1;
+            let waker = unsafe { std::ptr::read(self.inner[self.curr].as_mut_ptr()) };
+            waker.wake();
+        }
```

Comment:
Take it or leave it: this will generate a bounds check on every array access. I believe if we took a slice of the array and iterated over it, instead, we could potentially elide those bounds checks.

For example, we could write this, instead:
```suggestion
        for waker in self.inner[..self.curr].rev() {
            let waker = unsafe { std::ptr::read(waker.as_mut_ptr()) };
            waker.wake();
        }
```
Now, the compiler would hopefully generate only a single bounds check when `self.inner` is sliced, rather than every time the array is indexed.

Reply:
Update: I was actually slightly wrong about this: `rustc` 1.54 with `-C opt-level=3` won't generate bounds checks on every loop iteration, either, _but only when the assertion at the top of the method body is present_. See [this godbolt for some extremely cool optimizer behavior](https://godbolt.org/z/b8nTq76xs). 

However, the `for` loop version still does generate a _slightly_ nicer loop body. Compare the following (from the compiler explorer link):

**`while` loop (with assertion):**
```assembly
.LBB0_3:
        mov     qword ptr [r14 + 512], r15
        mov     rdi, qword ptr [rbx - 8]
        mov     rax, qword ptr [rbx]
        call    qword ptr [rax + 8]
        add     rbx, -16
        add     r15, -1
        jb      .LBB0_3
```
**`for` loop:**
```assembly
.LBB2_3:
        mov     rdi, qword ptr [r14 + rbx - 16]
        mov     rax, qword ptr [r14 + rbx - 8]
        call    qword ptr [rax + 8]
        add     rbx, -16
        jne     .LBB2_3
```

Note that the `for` loop version generates two fewer instructions than the `while` loop version. Which, admittedly, is not _huge_ --- both loops are optimized fairly nicely. But, given that this code is hot enough that it seems like this kind of microoptimization matters, this may be worth considering.

Reply:
**Another update (sorry):** Upon thinking about it a bit further, I realized that the `for` loop version may actually be wrong. The way we generated a better loop body in that case was by moving the mutation of `self.curr` out of the loop, and only mutating it when the loop has completed. However, because we also use `self.curr` to determine the length of the slice in `WakeList::drop`, this is actually wrong. 

If we don't update `self.curr` on every iteration, and a call to `wake` panics, the `Drop` impl will see the same value of `self.curr` we had at the _beginning_ of the loop. So, in that case, the `Drop` impl will try to call `drop_in_place` on wakers that have already been `ptr::read` out of the array, and were dropped by their `Waker::wake` impl. In order to avoid dropping these wakers a second time, we actually _have_ to update `self.curr` on every iteration.

Reply:
Yes, I realized that after committing as well. It actually reproduces the initial bug with drop after panic.

Reply:
We could write a version where the `for` loop also decrements `self.curr` in the loop body...but that gives us almost exactly the same assembly as the `while` loop: https://godbolt.org/z/ddjfYPdns so it's only worth doing if you find the `for` loop more readable. 

Sorry for taking you on a bit of a wild goose chase with this review! :)

Reply:
Have you considered setting `self.curr` before the loop? This way you wouldn't have to update it inside the loop, it would leak memory on panic, but never execute a double-free.

Reply:
> Have you considered setting `self.curr` before the loop? This way you wouldn't have to update it inside the loop, it would leak memory on panic, but never execute a double-free.

Hmm, I'm not sure if it's a good idea to leak here, since Tokio catches panics in tasks. It's conceivable for a user to ignore the result of `JoinHandle.await`, or to treat panicked tasks as non-fatal (e.g. in implementing a BEAM-like system). So, I'm unsure if we want to potentially leak memory on panics, if we can't rely on them necessarily meaning an abort/exit is imminent.

But, OTTOH, it would definitely improve performance in a hot loop in the more common case, and _most_ users probably want panics to eventually lead to exits...